### ISSUE TITLE
About Gradient Explosion Issues

### ISSUE THREAD

**Issue by zstar239** [JotaBravo/spacecraft-uda#17]
- Why does my training always encounter gradient explosion issues and never succeed?

![Image](https://github.com/user-attachments/assets/11f9fcb0-b661-46a5-b378-cfbb9fa29ca8)

![Image](https://github.com/user-attachments/assets/c5b288f1-35bc-49dc-9bba-95387b62671a)

![Image](https://github.com/user-attachments/assets/29f33073-41b4-43d7-8892-f635352eb79b)

![Image](https://github.com/user-attachments/assets/18ed1490-e2dc-417e-a84a-5e5d5bece3cf)

- Because gradient explosion keeps occurring, I added a condition check and lowered the learning rate, but the problem still persists.

**Comment by flpeng2010** [comment]
Hi, I’ve encountered the same issue as well. May I ask if you’ve already resolved it?

**Comment by zstar239** [comment]
You can turn off the other two losses first and **only use the heatmap loss** training to see if this problem will most likely not occur. If you need to use the other two losses, you can try training a few epochs before using the others. I just guess this issue is caused by BPNP solution instability. Of course, I have made a simple modification of the BPNP method, and you can also try it out using this.
‘’‘

import torch
import numpy as np
import cv2 as cv
import kornia as kn

class BPnP_fast2(torch.autograd.Function):
    @staticmethod
    def forward(ctx, pts2d, pts3d, K, ini_pose=None):
        bs = pts2d.size(0)
        n = pts2d.size(1)
        device = pts2d.device
        pts3d_np = np.array(pts3d.detach().cpu())
        K_np = np.array(K.detach().cpu())
        P_6d = torch.zeros(bs,6,device=device, dtype=torch.float)

        # 缓存上一次有效的rvec和tvec，用于PnP失败时fallback
        last_rvec = None
        last_tvec = None

        for i in range(bs):
            pts2d_i_np = np.ascontiguousarray(pts2d[i].detach().cpu().numpy().reshape((n,1,2)))

            # solvePnPRansac返回值检测
            success, rvec0, tvec0, inliers = cv.solvePnPRansac(
                objectPoints=pts3d_np,
                imagePoints=pts2d_i_np,
                cameraMatrix=K_np,
                distCoeffs=None,
                flags=cv.SOLVEPNP_ITERATIVE,
                confidence=0.9999,
                reprojectionError=3)

            if not success:
                # fallback，使用上一次有效值或者初始化为零向量
                if last_rvec is not None and last_tvec is not None:
                    rvec0, tvec0 = last_rvec, last_tvec
                else:
                    rvec0 = np.zeros((3,1), dtype=np.float32)
                    tvec0 = np.zeros((3,1), dtype=np.float32)
            else:
                last_rvec, last_tvec = rvec0, tvec0

            if ini_pose is not None:
                rvec0 = ini_pose[i, 0:3].detach().cpu().numpy().reshape(3,1)
                tvec0 = ini_pose[i, 3:6].detach().cpu().numpy().reshape(3,1)

            # solvePnP用初始猜测
            success2, rvec, tvec = cv.solvePnP(
                objectPoints=pts3d_np,
                imagePoints=pts2d_i_np,
                cameraMatrix=K_np,
                distCoeffs=None,
                flags=cv.SOLVEPNP_ITERATIVE,
                useExtrinsicGuess=True,
                rvec=rvec0,
                tvec=tvec0)

            if not success2:
                # fallback 同上
                if last_rvec is not None and last_tvec is not None:
                    rvec, tvec = last_rvec, last_tvec
                else:
                    rvec = np.zeros((3,1), dtype=np.float32)
                    tvec = np.zeros((3,1), dtype=np.float32)

            angle_axis = torch.tensor(rvec, device=device, dtype=torch.float32).view(1,3)
            T = torch.tensor(tvec, device=device, dtype=torch.float32).view(1,3)
            P_6d[i,:] = torch.cat((angle_axis, T), dim=-1)

        ctx.save_for_backward(pts2d, P_6d, pts3d, K)
        return P_6d

    @staticmethod
    def backward(ctx, grad_output):
        pts2d, P_6d, pts3d, K = ctx.saved_tensors
        device = pts2d.device
        bs = pts2d.size(0)
        n = pts2d.size(1)
        m = 6

        grad_x = torch.zeros_like(pts2d)
        grad_z = torch.zeros_like(pts3d)
        grad_K = torch.zeros_like(K)

        lambda_eye = 1e-6 * torch.eye(m, device=device)

        for i in range(bs):
            J_fy = torch.zeros(m,m, device=device)
            J_fx = torch.zeros(m,2*n, device=device)
            J_fz = torch.zeros(m,3*n, device=device)
            J_fK = torch.zeros(m,9, device=device)

            coefs = get_coefs(P_6d[i].view(1,6), pts3d, K, create_graph=False).detach()

            pts2d_flat = pts2d[i].clone().view(-1).detach().requires_grad_()
            P_6d_flat = P_6d[i].clone().view(-1).detach().requires_grad_()
            pts3d_flat = pts3d.clone().view(-1).detach().requires_grad_()
            K_flat = K.clone().view(-1).detach().requires_grad_()

            for j in range(m):
                torch.set_grad_enabled(True)
                if j > 0:
                    pts2d_flat.grad.zero_()
                    P_6d_flat.grad.zero_()
                    pts3d_flat.grad.zero_()
                    K_flat.grad.zero_()

                R = kn.geometry.angle_axis_to_rotation_matrix(P_6d_flat[0:m-3].view(1,3))

                P = torch.cat((R[0,0:3,0:3].view(3,3), P_6d_flat[m-3:m].view(3,1)), dim=-1)
                KP = torch.mm(K_flat.view(3,3), P)
                pts2d_i = pts2d_flat.view(n,2).transpose(0,1)
                pts3d_i = torch.cat((pts3d_flat.view(n,3),torch.ones(n,1,device=device)), dim=-1).t()
                proj_i = KP.mm(pts3d_i)
                Si = proj_i[2,:].view(1,n)

                r = pts2d_i*Si - proj_i[0:2,:]
                coef = coefs[:,:,j].transpose(0,1) # size: [2,n]
                fj = (coef*r).sum()
                fj.backward()

                J_fy[j,:] = P_6d_flat.grad.clone()
                J_fx[j,:] = pts2d_flat.grad.clone()
                J_fz[j,:] = pts3d_flat.grad.clone()
                J_fK[j,:] = K_flat.grad.clone()

            # 正则化避免奇异
            J_fy_reg = J_fy + lambda_eye

            # 最小二乘求解雅可比逆
            J_yx = (-1) * torch.linalg.lstsq(J_fy_reg, J_fx).solution
            J_yz = (-1) * torch.linalg.lstsq(J_fy_reg, J_fz).solution
            J_yK = (-1) * torch.linalg.lstsq(J_fy_reg, J_fK).solution

            grad_x[i] = grad_output[i].view(1,m).mm(J_yx).view(n,2)
            grad_z += grad_output[i].view(1,m).mm(J_yz).view(n,3)
            grad_K += grad_output[i].view(1,m).mm(J_yK).view(3,3)

        return grad_x, grad_z, grad_K, None

’‘’

**Comment by zstar239** [comment]
You can also try to reduce the learning rate, which may alleviate this problem. But I haven't tried it, I can't guarantee it will work.

**Comment by flpeng2010** [comment]
> You can turn off the other two losses first and **only use the heatmap loss** training to see if this problem will most likely not occur. If you need to use the other two losses, you can try training a few epochs before using the others. I just guess this issue is caused by BPNP solution instability. Of course, I have made a simple modification of the BPNP method, and you can also try it out using this. ‘’‘
> 
> import torch import numpy as np import cv2 as cv import kornia as kn
> 
> class BPnP_fast2(torch.autograd.Function): [@staticmethod](https://github.com/staticmethod) def forward(ctx, pts2d, pts3d, K, ini_pose=None): bs = pts2d.size(0) n = pts2d.size(1) device = pts2d.device pts3d_np = np.array(pts3d.detach().cpu()) K_np = np.array(K.detach().cpu()) P_6d = torch.zeros(bs,6,device=device, dtype=torch.float)
> 
> ```
>     # 缓存上一次有效的rvec和tvec，用于PnP失败时fallback
>     last_rvec = None
>     last_tvec = None
> 
>     for i in range(bs):
>         pts2d_i_np = np.ascontiguousarray(pts2d[i].detach().cpu().numpy().reshape((n,1,2)))
> 
>         # solvePnPRansac返回值检测
>         success, rvec0, tvec0, inliers = cv.solvePnPRansac(
>             objectPoints=pts3d_np,
>             imagePoints=pts2d_i_np,
>             cameraMatrix=K_np,
>             distCoeffs=None,
>             flags=cv.SOLVEPNP_ITERATIVE,
>             confidence=0.9999,
>             reprojectionError=3)
> 
>         if not success:
>             # fallback，使用上一次有效值或者初始化为零向量
>             if last_rvec is not None and last_tvec is not None:
>                 rvec0, tvec0 = last_rvec, last_tvec
>             else:
>                 rvec0 = np.zeros((3,1), dtype=np.float32)
>                 tvec0 = np.zeros((3,1), dtype=np.float32)
>         else:
>             last_rvec, last_tvec = rvec0, tvec0
> 
>         if ini_pose is not None:
>             rvec0 = ini_pose[i, 0:3].detach().cpu().numpy().reshape(3,1)
>             tvec0 = ini_pose[i, 3:6].detach().cpu().numpy().reshape(3,1)
> 
>         # solvePnP用初始猜测
>         success2, rvec, tvec = cv.solvePnP(
>             objectPoints=pts3d_np,
>             imagePoints=pts2d_i_np,
>             cameraMatrix=K_np,
>             distCoeffs=None,
>             flags=cv.SOLVEPNP_ITERATIVE,
>             useExtrinsicGuess=True,
>             rvec=rvec0,
>             tvec=tvec0)
> 
>         if not success2:
>             # fallback 同上
>             if last_rvec is not None and last_tvec is not None:
>                 rvec, tvec = last_rvec, last_tvec
>             else:
>                 rvec = np.zeros((3,1), dtype=np.float32)
>                 tvec = np.zeros((3,1), dtype=np.float32)
> 
>         angle_axis = torch.tensor(rvec, device=device, dtype=torch.float32).view(1,3)
>         T = torch.tensor(tvec, device=device, dtype=torch.float32).view(1,3)
>         P_6d[i,:] = torch.cat((angle_axis, T), dim=-1)
> 
>     ctx.save_for_backward(pts2d, P_6d, pts3d, K)
>     return P_6d
> 
> @staticmethod
> def backward(ctx, grad_output):
>     pts2d, P_6d, pts3d, K = ctx.saved_tensors
>     device = pts2d.device
>     bs = pts2d.size(0)
>     n = pts2d.size(1)
>     m = 6
> 
>     grad_x = torch.zeros_like(pts2d)
>     grad_z = torch.zeros_like(pts3d)
>     grad_K = torch.zeros_like(K)
> 
>     lambda_eye = 1e-6 * torch.eye(m, device=device)
> 
>     for i in range(bs):
>         J_fy = torch.zeros(m,m, device=device)
>         J_fx = torch.zeros(m,2*n, device=device)
>         J_fz = torch.zeros(m,3*n, device=device)
>         J_fK = torch.zeros(m,9, device=device)
> 
>         coefs = get_coefs(P_6d[i].view(1,6), pts3d, K, create_graph=False).detach()
> 
>         pts2d_flat = pts2d[i].clone().view(-1).detach().requires_grad_()
>         P_6d_flat = P_6d[i].clone().view(-1).detach().requires_grad_()
>         pts3d_flat = pts3d.clone().view(-1).detach().requires_grad_()
>         K_flat = K.clone().view(-1).detach().requires_grad_()
> 
>         for j in range(m):
>             torch.set_grad_enabled(True)
>             if j > 0:
>                 pts2d_flat.grad.zero_()
>                 P_6d_flat.grad.zero_()
>                 pts3d_flat.grad.zero_()
>                 K_flat.grad.zero_()
> 
>             R = kn.geometry.angle_axis_to_rotation_matrix(P_6d_flat[0:m-3].view(1,3))
> 
>             P = torch.cat((R[0,0:3,0:3].view(3,3), P_6d_flat[m-3:m].view(3,1)), dim=-1)
>             KP = torch.mm(K_flat.view(3,3), P)
>             pts2d_i = pts2d_flat.view(n,2).transpose(0,1)
>             pts3d_i = torch.cat((pts3d_flat.view(n,3),torch.ones(n,1,device=device)), dim=-1).t()
>             proj_i = KP.mm(pts3d_i)
>             Si = proj_i[2,:].view(1,n)
> 
>             r = pts2d_i*Si - proj_i[0:2,:]
>             coef = coefs[:,:,j].transpose(0,1) # size: [2,n]
>             fj = (coef*r).sum()
>             fj.backward()
> 
>             J_fy[j,:] = P_6d_flat.grad.clone()
>             J_fx[j,:] = pts2d_flat.grad.clone()
>             J_fz[j,:] = pts3d_flat.grad.clone()
>             J_fK[j,:] = K_flat.grad.clone()
> 
>         # 正则化避免奇异
>         J_fy_reg = J_fy + lambda_eye
> 
>         # 最小二乘求解雅可比逆
>         J_yx = (-1) * torch.linalg.lstsq(J_fy_reg, J_fx).solution
>         J_yz = (-1) * torch.linalg.lstsq(J_fy_reg, J_fz).solution
>         J_yK = (-1) * torch.linalg.lstsq(J_fy_reg, J_fK).solution
> 
>         grad_x[i] = grad_output[i].view(1,m).mm(J_yx).view(n,2)
>         grad_z += grad_output[i].view(1,m).mm(J_yz).view(n,3)
>         grad_K += grad_output[i].view(1,m).mm(J_yK).view(3,3)
> 
>     return grad_x, grad_z, grad_K, None
> ```
> 
> ’‘’

Thank you very much! I applied the changes you suggested, and it seems to be working now.

I have one more question : it seems that the backward function of the BPnP_fast is never called? I added a breakpoint there, but it has never been triggered.

**Comment by zstar239** [comment]
<img width="841" height="909" alt="Image" src="https://github.com/user-attachments/assets/2622fd77-f4c8-4850-9631-df1fe2ccec29" />

I tested the issue you mentioned about breakpoints not triggering backpropagation, and it seems to exist. I searched on Google, and it seems that custom backpropagation doesn't support debugging; if you need debugging, you can use a third-party library. I don't know if that's really the reason. If you're interested, feel free to explore further. I used a simple print statement in the bpnp backpropagation method to show that it has entered backpropagation.
[https://blog.csdn.net/qq_41683065/article/details/123551427](https://blog.csdn.net/qq_41683065/article/details/123551427)

**Comment by flpeng2010** [comment]
Thanks a lot！
